File: app/convnet/services/layers/OutputLayer.ts
import {Layer} from '../Layer'
import {Utils} from '../utils'

export class OutputLayer extends Layer {
    private W:number[][];
    private deltas:number[];

    constructor(params:any) {
        super(params);
        this.deltas = [];
        this.initWeights();
    }

    private initWeights() {
        let prevSize = this.prev.getSize();
        this.W = [];
        for (let i=0; i<prevSize[0]; i++) {
            this.W[i] = [];

            for (let j=0; j<this.size[0]; j++) {
                this.W[i][j] = Utils.getRandom(-1, 1);
            }
        }
    }

    public feadforward() {
        let prevData = this.prev.getOutput();

        for (let j=0; j<this.size[0]; j++) {
            let S = 0;
            for (let i=0; i<this.prev.getSize()[0];i++) {
                S+=prevData.out[i]*this.W[i][j];
            }
            this.in[j] = S;
            this.out[j] = Utils.sigmoid(S);
        }

        this.net.setOutput(this.out);
    }

    public backprop() {
        let label = this.net.getLabel();
        let prevConfig = this.prev.getConfig();
        let prevOutput = this.prev.getOutput()['out'];
        
        for (let i=0; i<this.size[0]; i++) {
            this.deltas[i] = label[i] - this.out[i];
        }

        for (let j=0; j<this.size[0]; j++) {
            for (let i=0; i<prevConfig['size'][0];i++) {
                this.W[i][j]+=0.001*this.deltas[j]*Utils.sigmoidDerivative(this.in[j])*prevOutput[i];
            }
        }
    }

    public getOutput() {
        return {out: this.out};
    }

    public getDeltas() {
        return this.deltas;
    }

    public getWeights() {
        return this.W;
    }

    public getInput() {
        return this.in;
    }

    public getError() {
        return this.deltas.reduce((prev,current) => {
            return Math.pow(prev,2)+Math.pow(current,2);
        })/2;
    }
}

File: api/convnet/layers/OutputLayer.go
package layers

import (
	//"fmt"
	"math"
	"math/rand"
	. "github.com/alexkarpovich/convnet/api/convnet/utils"
)

type OutputLayer struct {
	*Layer
	weights [][]float64
	deltas []float64
}

func (l *OutputLayer) Prepare() {
	inSize := l.prev.GetProp("outSize").([]int)
	length := inSize[0]

	l.weights = make([][]float64, length)
	for i:=0; i<length; i++ {
		l.weights[i] = make([]float64, l.size[0])

		for j:=0; j<l.size[0]; j++ {
			l.weights[i][j] = rand.Float64()*2-1
		}
	}

	l.in = make([]float64, l.size[0])
	l.out = make([]float64, l.size[0])
	l.deltas = make([]float64, l.size[0])
}

func (l *OutputLayer) FeedForward() {
	prevOut := l.prev.GetProp("out").([]float64)
	prevSize := l.prev.GetProp("outSize").([]int)

	for j:=0; j<l.size[0]; j++ {
		s := 0.0

		for i:=0; i<prevSize[0]; i++ {
			s += prevOut[i] * l.weights[i][j]
		}

		l.in[j] = s
		l.out[j] = Sigmoid(s)
	}

	l.net.SetOutput(l.out)
	l.net.SetError(l.GetError())
}

func (l *OutputLayer) BackProp() {
	alpha := 0.001
	label := l.net.GetLabel();
	inSize := l.prev.GetProp("outSize").([]int)
	prevOut := l.prev.GetProp("out").([]float64)

	for i:=0; i<l.size[0]; i++ {
		l.deltas[i] = label[i] - l.out[i];
	}

	for j:=0; j<l.size[0]; j++ {
		for i:=0; i<inSize[0];i++ {
			l.weights[i][j] += alpha*l.deltas[j]*DSigmoid(l.in[j])*prevOut[i];
		}
	}
}

func (l *OutputLayer) GetProp(name string) interface{} {
	switch name {
	case "outSize": return l.size
	case "in": return l.in
	case "deltas": return l.deltas
	case "weights": return l.weights
	}

	return nil
}

func (l *OutputLayer) GetError() float64 {
	err := 0.0

	for i := range l.deltas {
		err += math.Pow(l.deltas[i], 2)
	}

	return err
}
